File: mobile/wear/app/src/main/java/sh/martinez/pal/companion/wear/PALClient.kt
package sh.martinez.pal.companion.wear

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Minimal PAL API client for the watch: one call, `command`, which POSTs the
 * dictated text with wait_reply=true so PAL's reply comes back in the HTTP
 * response (the watch scope has no /ws/ui channel for async replies). Talks to
 * the gateway base over HTTPS — the same path as away-from-home use.
 */
object PALClient {

    class Failure(message: String) : Exception(message)

    /** Send a command, return PAL's reply text ("" if none). */
    suspend fun command(text: String): String = withContext(Dispatchers.IO) {
        val url = URL("${DevConfig.SERVER_BASE}/api/command")
        val conn = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            connectTimeout = 10_000
            readTimeout = 95_000          // server caps the turn at 90s
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("Authorization", "Bearer ${DevConfig.WATCH_TOKEN}")
        }
        try {
            val body = JSONObject()
                .put("text", text)
                .put("wait_reply", true)
                .toString()
            conn.outputStream.use { it.write(body.toByteArray()) }

            val code = conn.responseCode
            if (code != 200) throw Failure("HTTP $code")
            val resp = conn.inputStream.bufferedReader().use { it.readText() }
            JSONObject(resp).optString("reply", "").trim()
        } finally {
            conn.disconnect()
        }
    }
}

File: mobile/ios/App/PALWatch/PALClient.swift
import Foundation

/// Minimal PAL API client for the watch: one call, `command`, which POSTs the
/// dictated text with wait_reply=true so PAL's reply comes back in the HTTP
/// response (the watch scope has no /ws/ui channel for async replies). Talks
/// to the gateway base over HTTPS — same path as away-from-home use.
enum PALClient {
    struct CommandRequest: Encodable {
        let text: String
        let wait_reply: Bool
    }

    struct CommandResponse: Decodable {
        let status: String
        let reply: String?
        let message: String?
    }

    enum Failure: LocalizedError {
        case httpStatus(Int)
        case badPayload
        case notPaired

        var errorDescription: String? {
            switch self {
            case .httpStatus(let code): return "HTTP \(code)"
            case .badPayload: return "unexpected response"
            case .notPaired: return "not paired"
            }
        }
    }

    /// Send a command and return PAL's spoken reply text ("" if none).
    static func command(_ text: String) async throws -> String {
        guard let base = ConfigStore.base, let token = ConfigStore.token,
              !base.isEmpty, !token.isEmpty,
              let url = URL(string: "\(base)/api/command") else {
            throw Failure.notPaired
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        // The turn can take a while (LLM + tools); server caps at 90s.
        request.timeoutInterval = 95
        request.httpBody = try JSONEncoder().encode(
            CommandRequest(text: text, wait_reply: true))

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw Failure.badPayload
        }
        guard http.statusCode == 200 else {
            throw Failure.httpStatus(http.statusCode)
        }
        let decoded = try JSONDecoder().decode(CommandResponse.self, from: data)
        return (decoded.reply ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
